feat: Proxies for server 4 (#7041)
circleci/circleci-docs: jekyll/_cci2/server/operator/installing-server-behind-a-proxy.adoc

[#installation-and-configuration]
== Installation and configuration

To enable the proxy, first toggle `proxy.enabled` to `"1"`.

Next, fill out `proxy.http.host` and `proxy.https.host`, along with their associated port. These values can be the same but they both need to be configured.

In order to configure authentication, you will need to configure `proxy.http.auth.enabled` and `proxy.https.auth.enabled` as `"1"`. You will also need to configure the respective username and password for both HTTP and HTTPS.

Finally, configure the `no_proxy` hosts and subnets. This should include localhost, your GitHub Enterprise host (optional), and the hostname of your CircleCI installation (see limitations below for an explanation).

```yaml
proxy:
  enabled: "1"
  http:
    host: proxy.example.internal
    port: "3128"
    auth:
      enabled: "1"

>       username: circleci
>       password: circleciforever

Reviewer: ```suggestion
      username: <proxy-user>
      password: <proxy-password>
```

  https:
    host: proxy.example.internal
    port: "3128"
    auth:
      enabled: "1"
      username: circleci
      password: circleciforever
  no_proxy:
    - localhost
    - 127.0.0.1
    - github.example.internal
    - circleci.example.internal

>     - 10.0.0.0/16 # Subnets to exclude from the proxy (optional)

Reviewer: ```suggestion
    - <vpc-or-subnet-cidr>   # VPC or Subnets to exclude from the proxy (optional)
```

>     - <vpc-or-subnet-cidr>   # VPC or subnets to exclude from the proxy (optional)

Reviewer 2: Should we make a suggestion to put nomad subnets in here, or are those covered by something else?

```

[#known-limitations]
== Known limitations

* The CircleCI hostname must be added to the no-proxy list because the following services: `output processor` and `vm-service`. The application and build-agent share a no-proxy list and are assumed to be behind the same firewall and therefore cannot have a proxy between them.
* Some additional configuration is required to import orbs when installed behind a proxy. See https://circleci.com/docs/2.0/server-3-operator-orbs/#using-orbs-behind-a-proxy[Orbs on Server] docs for more information.
* The JVM only accepts proxies that run over HTTP, not HTTPS, and therefore proxy URIs must be of the form `\http://user:password@host:port` rather than `\https://user:password@host:port`.

> * If your GitHub instance is running outside of the proxied environment (either GitHub.com or GitHub Enterprise), you must ensure that SSH traffic from our application (inside the Kubernetes cluster) and from our Nomad node can reach your instance without additional configuration. Our SSH agents do not respect proxy settings because it uses a different network protocol. If a proxy is the only way that any traffic can reach outside the proxied environment, this means it will block SSH traffic and your application will fail.

Author: This bullet point needs help. We cannot proxy ssh-keyscan but we can technically proxy SSH itself.

> * If your GitHub instance is running outside of the proxied environment (either GitHub.com or GitHub Enterprise), you must ensure that SSH traffic from our application (inside the Kubernetes cluster) and from our Nomad node can reach your instance. Please note the default `checkout` step in a CircleCI job will fail to clone code and our `ssh-keyscan` of GitHub Enterprise will not work. While you may configure an SSH proxy, `ssh-keyscan` can NOT be proxied and instead will require you provide `github.fingerprint` when using GHE.

Reviewer 2: NIT:
```suggestion
* If your GitHub instance is running outside of the proxied environment (either GitHub.com or GitHub Enterprise), you must ensure that SSH traffic from CircleCI (inside the Kubernetes cluster) and from our Nomad node can reach your instance. Please note the default `checkout` step in a CircleCI job will fail to clone code and our `ssh-keyscan` of GitHub Enterprise will not work. While you may configure an SSH proxy, `ssh-keyscan` can NOT be proxied and instead will require you provide `github.fingerprint` when using GHE.
```
